### PR TITLE
Towards a faster (sum) reduce

### PR DESCRIPTION
**Pull request by marcelbischoff**
I looked into an optimized sum reduce following https://dournac.org/info/gpu_sum_reduction 

* This implementation uses local memory and does reduction by a size `sz` up to 256 using the above algorithm.
* If the reduction size `sz >256` it computes partial sums of `groups ~ sz/256` working groups and creates an intermediate output of size (osize, groups) which is reduced recursively by summing over `groups`
* It gets fast for long reductions for example: `testing                    [100000000]   torch/tinygrad fp: 40.46 / 0.94 ms` that same test crashes after 5minutes on the current implementation
* This looses most of the ability of modifying the extra code pieces. Some can likely be built in but it is only ever used in `logsoftmax` which I think should really be a second kind op.

Just want to get an opinion

ToDo's
* needs to be cleaned up a lot but passes test as far as I can see.
* <del>breaks backward in serious_mnist and there is a bug for general sums if `sz>256`</del>
* needs better test<del>, definitely does not work in general</del>
* loop over recursion?

### CODE REVIEW

**Comment by marcelbischoff** [comment]
I agree that Convs are most important but I think once they are fixed all the unbroadcast will be a major slowdown. I think unbroadcast is not tracked, separately. I was going by complexity Sum, MatMul, Conv.
 
A lot are just definitions of variables and it is complex because of the axis support. Will looking into streamlining and benchmarking. If we just wanted full reduce that would be very short. I think unbroadcast is not tracked, separately. Most of the runtime is python overhead, I think. It scales very well because every kernel just one fp global memory access and less than a dozen 8-9 l1 cache and fp operations.

File: tinygrad/ops_gpu.py [hunk 1]
```diff
@@ -276,6 +303,25 @@ def backward(ctx, grad_output):
...
+    if axis is not None:
+      ret.shape = tuple([input.shape[i] for i in range(len(input.shape)) if i not in axis])
+    return ret
+
+  @staticmethod
+  def backward(ctx, grad_output):
+    input, axis, ret = ctx.saved_tensors
+    shape = [1 if axis is None or i in axis else input.shape[i] for i in range(len(input.shape))]
+    #output = GPUBuffer(shape, hostbuf=grad_output)
+    ret2 = binary_op(ctx, "1.0*(a == b)", input, ret)
```
Comment (geohot) [inline tinygrad/ops_gpu.py:321]: So what do you do in the case of ties?

**Comment by geohot** [comment]
Merged parts of this https://github.com/geohot/tinygrad/commit/27208d729ba13de5ea3abedad5cbdbf56adffee2

**Comment by marcelbischoff** [comment]
The partial derivatives don’t technically exist. For f(x,y)=max(x,y) the gradient will be (1,1) on the diagonal. argmax will give (1,0)?!

**Comment by geohot** [comment]
Hmm, we should probably have CPU match this too and add a test.

**Comment by marcelbischoff** [comment]
I was going to test what torch does.

**Comment by marcelbischoff** [comment]
Torch gives (.5,.5) as gradient. So we have to divide by `sum(ret2)`.

**Comment by geohot** [comment]
I wrote a faster one in "risk" I think, really we should refactor the bottom layer out.

**Comment by geohot** [comment]
Ideally this sort of stuff should be in a middleware somewhere.

**Pull request closed without merging**